driver-core/src/main/com/mongodb/internal/connection/IdTrackingBsonWriter.java

```diff
@@ -0,0 +1,443 @@
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import org.bson.BsonBinary;
+import org.bson.BsonBinaryWriter;
+import org.bson.BsonBoolean;
+import org.bson.BsonDateTime;
+import org.bson.BsonDbPointer;
+import org.bson.BsonDecimal128;
+import org.bson.BsonDocument;
+import org.bson.BsonDouble;
+import org.bson.BsonInt32;
+import org.bson.BsonInt64;
+import org.bson.BsonJavaScript;
+import org.bson.BsonJavaScriptWithScope;
+import org.bson.BsonMaxKey;
+import org.bson.BsonMinKey;
+import org.bson.BsonNull;
+import org.bson.BsonObjectId;
+import org.bson.BsonReader;
+import org.bson.BsonRegularExpression;
+import org.bson.BsonString;
+import org.bson.BsonSymbol;
+import org.bson.BsonTimestamp;
+import org.bson.BsonUndefined;
+import org.bson.BsonValue;
+import org.bson.BsonWriter;
+import org.bson.RawBsonDocument;
+import org.bson.io.BasicOutputBuffer;
+import org.bson.types.Decimal128;
+import org.bson.types.ObjectId;
+
+import java.util.function.Supplier;
+
+public class IdTrackingBsonWriter extends BsonWriterDecorator {
+
+    private final LevelCountingBsonWriter idBsonBinaryWriter;
+    private static final String ID_FIELD_NAME = "_id";
+    private int level = -1;
+    private String currentFieldName;
+    private BsonValue id;
+
+    public IdTrackingBsonWriter(final BsonWriter bsonWriter) {
+        super(bsonWriter);
+        idBsonBinaryWriter = new LevelCountingBsonWriter(new BsonBinaryWriter(new BasicOutputBuffer())){};
+    }
+
+    @Override
+    public void writeStartDocument(final String name) {
+        setCurrentFieldName(name);
+
+        if (isWritingId()) {
+            idBsonBinaryWriter.writeStartDocument(name);
+        }
+        level++;
+        super.writeStartDocument(name);
+    }
+
+    @Override
+    public void writeStartDocument() {
+        if (isWritingId()) {
+            idBsonBinaryWriter.writeStartDocument();
+        }
+        level++;
+        super.writeStartDocument();
+    }
+
+    @Override
+    public void writeEndDocument() {
+        if (isWritingId()) {
+            if (idBsonBinaryWriter.getCurrentLevel() >= 0) {
+                idBsonBinaryWriter.writeEndDocument();
+            }
+            if (idBsonBinaryWriter.getCurrentLevel() == -1) {
+                if (id != null && id.isJavaScriptWithScope()) {
+                    id = new BsonJavaScriptWithScope(id.asJavaScriptWithScope().getCode(), new RawBsonDocument(getBytes()));
+                } else if (id == null) {
+                    id = new RawBsonDocument(getBytes());
+                }
+            }
+        }
+
+
+        if (level == 0 && id == null) {
+            id = new BsonObjectId();
+            writeObjectId(ID_FIELD_NAME, id.asObjectId().getValue());
+        }
+
+        level--;
+        super.writeEndDocument();
+    }
+
+    @Override
+    public void writeStartArray() {
+        if (isWritingId()) {
+            if (idBsonBinaryWriter.getCurrentLevel() == -1) {
```

Comment:
If writing an array - we have to nest the array inside a document when using BsonWriter, otherwise the validation will fail. Happily we can get the array back later as a `RawBsonArray` once the array is closed.